### DISCUSSION TITLE
StartUpAtBoot

### DISCUSSION THREAD

**Discussion by radlik01** [DelphiWorlds/Kastri#169] [Q&A]
Dear all,

I try to implement startUpAtBoot to my Android application but I receive the error on startup
java.lang.RuntimeException: Unable to start receiver com.delphiworlds.kastri.DWMultiBroadcastReceiver: java.lang.NullPointerException: Attempt to invoke virtual method 'int android.content.Intent.getFlags()' on a null object reference
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	at android.app.ActivityThread.handleReceiver(ActivityThread.java:3395)
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	at android.app.ActivityThread.-wrap17(Unknown Source:0)
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1860)
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:166)
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6861)
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:450)
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
05-23 15:31:36.768 23812 23812 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'int android.content.Intent.getFlags()' on a null object reference
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	at android.app.ContextImpl.startActivity(ContextImpl.java:875)
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	at android.app.ContextImpl.startActivity(ContextImpl.java:859)
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	at android.content.ContextWrapper.startActivity(ContextWrapper.java:372)
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	at android.content.ContextWrapper.startActivity(ContextWrapper.java:372)
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	at com.delphiworlds.kastri.DWMultiBroadcastReceiver.startApp(DWMultiBroadcastReceiver.java:129)
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	at com.delphiworlds.kastri.DWMultiBroadcastReceiver.checkStartupIntent(DWMultiBroadcastReceiver.java:178)
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	at com.delphiworlds.kastri.DWMultiBroadcastReceiver.onReceive(DWMultiBroadcastReceiver.java:344)
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	at android.app.ActivityThread.handleReceiver(ActivityThread.java:3385)
05-23 15:31:36.768 23812 23812 E AndroidRuntime: 	... 8 more

Any idea where is the problem ?

Thank you

Jan

**Comment by DelphiWorlds** [comment]
This line in the callstack:

```
com.delphiworlds.kastri.DWMultiBroadcastReceiver.checkStartupIntent(DWMultiBroadcastReceiver.java:178)
```

Indicates that it is executing this (Java) code, which is related to when the application is installed over a previous version:

```
    if (intent.getAction().equals(Intent.ACTION_MY_PACKAGE_REPLACED)) {
      if ((metaData != null) && metaData.containsKey(KEY_RESTART_AFTER_REPLACE)) {
        if (metaData.getBoolean(KEY_RESTART_AFTER_REPLACE))
          return startApp(context); // <----- This is line 178
      }
    }
```

Which is an odd thing to happen if the device is booting up. Can you provide a project or compiled .apk that demonstrates this issue?

**Comment by radlik01** [comment]
Here is a apk
[SIDSafeTracker.apk.zip](https://github.com/DelphiWorlds/Kastri/files/11551409/SIDSafeTracker.apk.zip)

**Comment by DelphiWorlds** [comment]
The manifest in your project is not right. You have 2 application entries when there should be only one. 

Are you using the `AndroidManifest.merge.xml` file with the Build Event [as per the readme](https://github.com/DelphiWorlds/Kastri/tree/master/Demos/AndroidStartAtBoot#readme)? 

If you are manually modifying `AndroidManifest.template.xml` instead, make sure you add the metadata and receiver entries to the *existing* `<application>` node, not a separate one.

**Comment by radlik01** [comment]
Yes i use merge
![image](https://github.com/DelphiWorlds/Kastri/assets/18271189/8d624520-2b34-4c02-9cc1-77a5393dd7ca)

**Comment by radlik01** [comment]
if I use the merge on command line I have got the same result
`manifestmerge AndroidManifest.merge.xml AndroidManifest.xml`
[AndroidManifest_after_merge.txt](https://github.com/DelphiWorlds/Kastri/files/11553307/AndroidManifest_after_merge.txt)
[AndroidManifest.txt](https://github.com/DelphiWorlds/Kastri/files/11553308/AndroidManifest.txt)
[AndroidManifest.merge.txt](https://github.com/DelphiWorlds/Kastri/files/11553309/AndroidManifest.merge.txt)

**Comment by radlik01** [comment]
I have merged the manifets manualy and it works

**Comment by DelphiWorlds** [comment]
I've pushed a fix for the merge file.. thanks!
